### PR TITLE
wait for cloud init & extract autoinstall config

### PR DESCRIPTION
**Pull request by mwhudson**
I guess I'd like to know if any of the cloud-init developers think this is insane...

### CODE REVIEW

File: subiquity/core.py [hunk 1]
```diff
@@ -151,6 +151,14 @@ def make_screen(self):
             return s
 
     def run(self):
+        if not os.path.exists(self.opts.autoinstall):
+            from cloudinit import atomic_helper, stages
+            cfg = stages.Init().cfg
+            if 'autoinstall' in cfg:
+                atomic_helper.write_file(
+                    self.opts.autoinstall,
+                    yaml.dump(cfg['autoinstall']),
```
Comment (raharper) [inline subiquity/core.py]: Unless your yaml is configured to use SafeDumper, you can use cloud-init's

from cloudinit import safeyaml
safeyaml.dumps(cfg)

File: subiquity/core.py [hunk 2]
```diff
@@ -151,6 +151,14 @@ def make_screen(self):
             return s
 
     def run(self):
+        if not os.path.exists(self.opts.autoinstall):
+            from cloudinit import atomic_helper, stages
+            cfg = stages.Init().cfg
```
Comment (raharper) [inline subiquity/core.py]: cfg = stages.Init().read_cfg()

cfg has some redactions, not clear to me if that will impact where you're looking.

Note the Init cfg property is /etc/cloud , /etc/cloud.cfg.d/* and datasource cfg, but it is NOT user-data.

I think what you want is something like this:
```
from cloudinit import stages
init = stages.Init()
init.read_cfg()
init.fetch(existing="trust")
cloud = init.cloudify()
full_config = cloud.cfg
```
This is the config that's passed to config modules and all of the system config, user-data, vendor-data have all be processed and merged (where processed means downloading from a URL, for example).  As an example, if the user-data were a multi-part mime, then you cannot read /var/lib/cloud/instance/user-data.txt directly (without doing more work)

```
>>> from cloudinit import stages
>>> init = stages.Init()
>>> init.read_cfg()
>>> init.fetch(existing='trust')
>>> cloud = init.cloudify()
>>> cloud.datasource.userdata
<email.mime.multipart.MIMEMultipart object at 0x7f1e4dc88760>
>>> cloud.cfg
{'final_message': 'The system is finally up, after $UPTIME seconds', 'package_update': True, 'package_upgrade': True, 'packages': ['ifupdown'], 'datasource_list': ['NoCloud', 'None'],
```

 

Now, if you from know where the autoinstall cfg is being supplied (for example the kernel cmdline is writing it to a file on /etc/cloud/cloud.cfg.d/ ... then we can make some short cuts here.

File: subiquity/cmd/tui.py [hunk 3]
```diff
@@ -117,6 +119,10 @@ def main():
     logger.info("Starting SUbiquity v{}".format(VERSION))
     logger.info("Arguments passed: {}".format(sys.argv))
 
+    ci_start = time.time()
+    run_command(["cloud-init", "status", "--wait"])
+    logger.debug("waited %ss for cloud-init", time.time() - ci_start)
+
```
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/cmd/tui.py]: I think the point is that we don't know, although I expect most use cases will be ds=nocloud-net;s=$url on the kernel command line or using a config drive.

Anyway, this wait is mostly for cosmetic reasons: it looks silly when subiquity starts and then cloud-init final sprays stuff over the console.

File: subiquity/core.py (hunk 1 above)
Comment (mwhudson) [reply subiquity/core.py]: OK sure, I didn't realize you needed to be safe when _dumping_ yaml!? (and in any case, the yaml contains shell scripts that are going to be executed as root so if someone controls your autoinstall data it's not like they have to be sneaky at all)

File: subiquity/cmd/tui.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (raharper) [reply subiquity/cmd/tui.py]: OK, that makes sense.

File: subiquity/core.py (hunk 2 above)
Comment (mwhudson) [reply subiquity/core.py]: > This is the config that's passed to config modules and all of the system config, user-data, vendor-data have all be processed and merged (where processed means downloading from a URL, for example).

great, that's exactly what I want

> As an example, if the user-data were a multi-part mime, then you cannot read /var/lib/cloud/instance/user-data.txt directly (without doing more work)

yeah, this is the sort of thing I was worried about.

> Now, if you from know where the autoinstall cfg is being supplied (for example the kernel cmdline is writing it to a file on /etc/cloud/cloud.cfg.d/ ... then we can make some short cuts here.

I would prefer not to make any assumptions here if I can help it.

File: subiquity/core.py (hunk 1 above)
Comment (raharper) [reply subiquity/core.py]: We've had at least one security review scanning code and scolding for not using safe loaders and dumpers.

**Comment by mwhudson** [comment]
I've pushed an update, please have a look when you can.

> Looking for more info on where/how autoinstall is fed into cloud-init; ie what will subiquity document on how to feed autoinstall to the live installer.

As I said in a comment somewhere, I expect people will use a config drive or nocloud-net, but I want to retain full flexibility.

**Pull request merged**